### PR TITLE
MOSIP-20534: Added validator for code and value fields of dynamic field

### PR DESCRIPTION
**Pull request by SwethaKrish4**

### CODE REVIEW

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/ValueValidator.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    @Override
+    public boolean isValid(JsonNode jsonValue, ConstraintValidatorContext context) {
+        JsonNode val=jsonValue.get("value");
+        String value= val.asText();
+        if(value!=null && !value.isEmpty()){
+            Pattern p= Pattern.compile(allowedCharactersRegex,Pattern.CASE_INSENSITIVE);
+            Matcher m=p.matcher(value.trim());
+            return !(m.find());
+        }
+        return true;
```
Comment (ase-101) [inline admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/ValueValidator.java:28]: This returns true when the value is empty.

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    @Override
+    public boolean isValid(JsonNode value, ConstraintValidatorContext context) {
+        JsonNode val=value.get("code");
+        String code= val.asText();
+        if(code!=null && !code.isEmpty()){
+            Pattern p= Pattern.compile(allowedCharactersRegex,Pattern.CASE_INSENSITIVE);
+            Matcher m=p.matcher(code.trim());
+            return !(m.find());
+        }
+        return true;
```
Comment (ase-101) [inline admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java:27]: This returns true when the code is empty.

File: admin/kernel-masterdata-service/src/test/java/io/mosip/kernel/masterdata/test/controller/DynamicFieldControllerTest.java [hunk 3]
```diff
@@ -131,18 +131,48 @@ public void t001createDynamicFieldTest() throws Exception {
...
+						mockMvc.perform(
+										MockMvcRequestBuilders.post("/dynamicfields").contentType(MediaType.APPLICATION_JSON)
+												.content(mapper.writeValueAsString(dynamicFieldDtoReq)))
+								.andReturn(),
+						"KER-MSD-999");
+	}
+
 	@Test
 	@WithUserDetails("global-admin")
 	public void t001createDynamicFieldTest1() throws Exception {
```
Comment (ase-101) [inline admin/kernel-masterdata-service/src/test/java/io/mosip/kernel/masterdata/test/controller/DynamicFieldControllerTest.java:167]: Let us follow the below convention when naming the test case
<methodNameBeingTested>_<testcase>_<testCaseStatus>
Eg: createDynamicField_withAllValidInput_thenPass
It clearly explains what is tested and its possible output.

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/ValueValidator.java (hunk 1 above)
Comment (SwethaKrish4) [reply admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/ValueValidator.java]: resolved

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java (hunk 2 above)
Comment (SwethaKrish4) [reply admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java]: resolved

File: admin/kernel-masterdata-service/src/test/java/io/mosip/kernel/masterdata/test/controller/DynamicFieldControllerTest.java (hunk 3 above)
Comment (SwethaKrish4) [reply admin/kernel-masterdata-service/src/test/java/io/mosip/kernel/masterdata/test/controller/DynamicFieldControllerTest.java:167]: changed the testcase names

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+
+    @Override
+    public boolean isValid(JsonNode value, ConstraintValidatorContext context) {
+        JsonNode val = value.get("code");
+        String code = val.asText();
+        if (code == null && code.isEmpty()) {
+            return false;
+        }
+        Pattern p = Pattern.compile(allowedCharactersRegex, Pattern.CASE_INSENSITIVE);
```
Comment (ase-101) [inline admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java:25]: Can we compile the regex only once and reuse the compiled regex?

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/dto/DynamicFieldDto.java [hunk 5]
```diff
@@ -47,5 +49,7 @@ public class DynamicFieldDto {
 	@NotBlank
 	private String description;
 
+	@DynamicCodeValidator
+	@DynamicValueValidator
```
Comment (ase-101) [inline admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/dto/DynamicFieldDto.java:53]: Why are we not validating code and value in single validator? is it required to have 2 different validators?

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+public class CodeValidator implements ConstraintValidator<DynamicCodeValidator, JsonNode> {
+    @Value("${mosip.kernel.masterdata.code.validate.regex}")
+    private String allowedCharactersRegex;
+
+
+    @Override
+    public boolean isValid(JsonNode value, ConstraintValidatorContext context) {
+        JsonNode val = value.get("code");
+        String code = val.asText();
+        if (code == null && code.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (SwethaKrish4) [reply admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java]: updated the code as per review comment

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java (hunk 4 above)
Comment (SwethaKrish4) [reply admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/validator/CodeValidator.java]: For now, I am compiling the regex just once.

File: admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/dto/DynamicFieldDto.java (hunk 5 above)
Comment (SwethaKrish4) [reply admin/kernel-masterdata-service/src/main/java/io/mosip/kernel/masterdata/dto/DynamicFieldDto.java]: removed two different validators and used the single validator for code and dynamic fields

**Review by dhanendra06 — APPROVED** [review]

**Review by ase-101 — APPROVED** [review]

**Pull request merged**